addons/mail/static/src/utils/messaging_component.js

```diff
@@ -0,0 +1,35 @@
+/** @odoo-module */
+
+import { useModels } from "@mail/component_hooks/use_models/use_models";
+import { useShouldUpdateBasedOnProps } from "@mail/component_hooks/use_should_update_based_on_props/use_should_update_based_on_props";
+
+const componentRegistry = {};
+
+/**
+ * Allows a component to lean on the messaging features: the component is
+ * subscribed to changes in messaging models in its environment, the component
+ * will only be rerenderd if its props change, and the component has access to
+ * all components that are registered with this function without having to
+ * explicitly import them.
+ *
+ * @param {function} ComponentClass the constructor of the component to be
+ *      registered. Its name will be used as its key in the registry.
+ * @param {Object} [param1]
+ * @param {Object} [param1.propsCompareDepth] @see useShouldUpdateBasedOnProps
+ */
+export function registerMessagingComponent(ComponentClass, { propsCompareDepth = {} } = {}) {
+    ComponentClass.components = Object.assign(Object.create(componentRegistry), ComponentClass.components);
+    const originalSetup = ComponentClass.prototype.setup;
+    ComponentClass.prototype.setup = function setup(...args) {
+        originalSetup.apply(this, args);
+        // useModels must be defined after useRenderedValues, indeed records and
+        // fields accessed during useRenderedValues should be observed by
+        // useModels as if they were part of the OWL rendering itself.
+        useModels();
+        useShouldUpdateBasedOnProps({ propsCompareDepth });
+    }
+    const key = ComponentClass.name;
+    const previousEntry = componentRegistry[key];
+    componentRegistry[key] = ComponentClass;
+    return () => componentRegistry[key] = previousEntry;
```

Comment:
If we actually want to handle a revert, we should probably have some kind of stack to avoid issues with interleaved calls...
Or just prevent overrides altogether for now. For the test I was thinking about a standalone removal method instead of a revert.

Reply:
The stack should be implemented in userspace. When using web's "registerCleanup", cleanups are already a stack and they're executed in reverse order from registration.

Reply:
I agree, but then I would do everything in user space. It's strange to partially handle the revert it, but with only one level of stack.
So here I would just stay very generic and provide way to:
- function to register a new component with a given name, with a crash if it already exists
- function to get the built/decorated component from a given name,
- function to unregister a component for a given name,
- (later if we actually need replace or override we should have a function to register them)